_posts/2025-04-25-from-recipes-to-vectors-building-a-hybrid-search-app-with-opensearch.md

```diff
@@ -0,0 +1,299 @@
+  "TotalTime": 265,
+  "Text": "
+      Biryani
+      Ingredients: saffron, milk...
+      Instructions: Soak saffron in warm...
+  ",
+  "Name": "Biryani",
+  "DatePublished": "1999-08-29T13:12:00Z"
+  "embedding": ["float_", [[0.03729248, 0.031402588, … ]]]	
+}
+```
+
+### Ingestion
+
+The final step in the data preparation process is ingesting the data into OpenSearch. With our index already set up, all that remains is to use OpenSearch Bulk API to insert the chunks along with their embeddings.
+
+During ingestion, OpenSearch will generate the HNSW graph based on the default configurations. Once the data is ingested, the index will be ready for vector-based queries.
+
+Execute the ingestion script to insert the documents into OpenSearch. Once completed, you can verify the stored data by fetching a document using the following command in Dev Tools:
+
+```
+POST /recipes/_search
+{
+   "size": 1,
+   "query": {
+       "match_all": {}
+   }
+}
+
+```
+
+This will return the stored document, allowing you to inspect its values, including the vector embedding.
+
+With the data successfully ingested, the next step is running queries to test and evaluate our hybrid search system! 
+
+## Searching the index
+
+Now that we’ve processed and ingested our data, we can move on to the most exciting part—searching the index. OpenSearch allows us to perform different types of searches, including keyword search, semantic search, and hybrid search.
+
+### Keyword Search
+
+Keyword search is the traditional method of retrieving documents based on exact matches or full-text relevance scoring. It works well for structured fields like the category, ingredients , so if I was craving chicken, I could search the index using a match query:
+
+```
+GET /recipes/_search
+{
+  "query": {
+    "match": {
+      "RecipeCategory": "Chicken"
+    }
+  }
+}
+
+-----
+Coronation Chicken Salad
+Greek Chicken and Rice
+Chicken Teriyaki with Cashew Pineapple Rice
+Chicken Noodle Soup
+Macaroni and Cheese a 'la King
+-----
+
+```
+
+### Semantic Search
+
+However, if I wanted to find chicken recipes that were asian inspired and featured noodles and peanuts, a simple keyword search wouldn’t be enough. Keyword-based retrieval relies on exact matches, meaning it would only return documents that explicitly contain those words—missing relevant results that describe the same concept in different terms.
+
+To retrieve the most relevant recipes based on meaning rather than exact wording, we need to use semantic search.
+
+First of all, we would need to pass our search query through the same embedding model (e.g., Cohere's `embed-english-v3`), so we can use the resulting vector as the vector in an OpenSearch knn query.
```

Comment:
```suggestion
First of all, we would need to pass our search query through the same embedding model (for example, Cohere's `embed-english-v3`) so that we can use the resulting vector as the vector in an OpenSearch k-NN query.
```